src/devices/SocketCan/CanFrame.cs

```diff
@@ -0,0 +1,145 @@
+using System.Threading;
+
+namespace Iot.Device.SocketCan
+{
+    [StructLayout(LayoutKind.Sequential)]
+    public unsafe struct CanFrame
+    {
+        private const int CAN_MAX_DLEN = 8;
+
+        // RawId (can_id) includes EFF, RTR and ERR flags
+        public uint RawId { get; set; }
+
+        // data length code (can_dlc)
+        // see: ISO 11898-1 Chapter 8.4.2.4
+        private byte _length;
+        private byte _pad;
+        private byte _res0;
+        private byte _res1;
+        private fixed byte _data[CAN_MAX_DLEN];
+
+        public uint Id => ExtendedFrameFormat ? ExtendedId : StandardId;
+
+        public uint StandardId
+        {
+            get
+            {
+                if (ExtendedFrameFormat)
+                    throw new InvalidOperationException($"{nameof(StandardId)} can be obtained only when {nameof(ExtendedFrameFormat)} is not set.");
+                
+                return RawId & Interop.CAN_SFF_MASK;
+            }
+            set
+            {
+                if ((value & ~Interop.CAN_SFF_MASK) != 0)
+                    throw new InvalidOperationException($"{nameof(value)} must be 11 bit identifier.");
+
+                ExtendedFrameFormat = false;
+                RawId &= ~Interop.CAN_SFF_MASK;
+                RawId |= value;
+            }
+        }
+
+        public uint ExtendedId
+        {
+            get
+            {
+                if (!ExtendedFrameFormat)
+                    throw new InvalidOperationException($"{nameof(ExtendedId)} can be obtained only when {nameof(ExtendedFrameFormat)} is set.");
+                
+                return RawId & Interop.CAN_EFF_MASK;
+            }
+            set
+            {
+                if ((value & ~Interop.CAN_EFF_MASK) != 0)
+                    throw new InvalidOperationException($"{nameof(value)} must be 29 bit identifier.");
+
+                ExtendedFrameFormat = true;
+                RawId &= ~Interop.CAN_EFF_MASK;
+                RawId |= value;
+            }
+        }
+
+        public bool Error
+        {
+            get => ((CanFlags)RawId).HasFlag(CanFlags.Error);
+            set => SetCanFlag(CanFlags.Error, value);
+        }
+
+        public bool ExtendedFrameFormat
+        {
+            get => ((CanFlags)RawId).HasFlag(CanFlags.ExtendedFrameFormat);
+            set => SetCanFlag(CanFlags.ExtendedFrameFormat, value);
+        }
+
+        public bool RemoteTransmissionRequest
+        {
+            get => ((CanFlags)RawId).HasFlag(CanFlags.RemoteTransmissionRequest);
+            set => SetCanFlag(CanFlags.RemoteTransmissionRequest, value);
+        }
+
+        public bool IsValid
+        {
+            get
+            {
+                uint idMask = ExtendedFrameFormat ? Interop.CAN_EFF_MASK : Interop.CAN_SFF_MASK;
+                return !Error && _length <= CAN_MAX_DLEN && (RawId & ~idMask) == 0;
+            }
+        }
+
+        public ReadOnlySpan<byte> Data
```

Comment:
As we spoke with @tarekgh offline I'll change this struct to ref struct to emphasize that struct is stack only

Reply:
You can't do this safely. You have to create an array here to be safe. I've been discussing a way to handle this safely in the future with @jaredpar.

If you've got a processing loop you can deal with this limitation the same way I did `FileSystemEntry` for file enumeration. It is a ref struct that you never directly create- it gets passed to your handlers in `FileSystemEnumerator`/`Enumerable` which makes it difficult to elevate the native view out of scope.